control_unit.py

```diff
@@ -22,22 +22,44 @@ def update_flags(reg_index):
 
 def _BR(instruction):
     """branch"""
-    pass
+    pc_offset = sign_extend((instruction) & 0x1ff, 9)
+    cond_flag = (instruction >> 9) & 0x7
+    if cond_flag & Registers.COND:
+        Registers.PC = ushort(Registers.PC + pc_offset)
 
 
 def _ADD(instruction):
     """add"""
-    pass
+    # destination register DR
+    DR = (instruction >> 9) & 0x7
+    # first operand SR1
+    SR1 = (instruction >> 6) & 0x7
+    # for immediate mode
+    imm_flag = (instruction >>5) & 0x1
+
+    if imm_flag:
+        imm5 = sign_extend(instruction & 0x1F, 5)
+        DR = ushort(Regisers.R0 + imm5) 
+    else:
+        SR2 = instruction & 0x7
+        DR = ushort(Registers.R0 + Registers.R1)
```

Comment:
same here, write to `Registers(DR)`, not the variable `DR` directly.
Also, You probably didn't mean to add `R0` contents to `R1` contents, instead, you need to add `SR1` contents to `SR2` contents.